File: backend/app/schemas/activity.py
"""Activity (SystemEvent) API schemas."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class ActivityEventSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    event_type: str
    severity: str
    source: str
    message: str
    actor: str | None
    correlation_id: str | None
    payload: dict | None
    occurred_at: datetime


class ActivityPageSchema(BaseModel):
    items: list[ActivityEventSchema]
    total: int
    page: int
    page_size: int

File: web/app/services/api/adapters/activity.ts
import type { ActivityEvent, ActivitySeverity } from '~/types/activity'
import type { Paginated } from '~/types/api'
import { resourceLinkFromPayload } from '~/utils/resourceLinks'
import { iso, normalizeEnvironment } from './common'
import type { RawActivityEvent, RawActivityPage } from './raw'

export function toActivityEvent(raw: RawActivityEvent): ActivityEvent {
  return {
    id: raw.id,
    component: raw.source,
    event_type: raw.event_type,
    severity: raw.severity as ActivitySeverity,
    message: raw.message,
    symbol: typeof raw.payload?.symbol === 'string' ? raw.payload.symbol : null,
    actor: raw.actor,
    correlation_id: raw.correlation_id,
    data: raw.payload,
    occurred_at: iso(raw.occurred_at),
    link: resourceLinkFromPayload(raw.payload),
    environment: normalizeEnvironment(raw.payload),
  }
}

export function toActivityPage(raw: RawActivityPage): Paginated<ActivityEvent> {
  return {
    items: raw.items.map(toActivityEvent),
    total: raw.total,
    page: raw.page,
    pageSize: raw.page_size,
  }
}
